avro-fastserde/src/main/java/com/linkedin/avro/primitive/PrimitiveGenericRecord.java

```diff
@@ -0,0 +1,103 @@
+package com.linkedin.avro.primitive;
+
+import java.nio.ByteBuffer;
+import org.apache.avro.AvroRuntimeException;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.util.Utf8;
+
+
+/**
+ * An extension to the {@link GenericRecord} API that includes utility functions for
+ * casting fields to the right type, including both boxed and unboxed. The following
+ * is worth keeping in mind:
+ *
+ * - All getter and setter functions can throw {@link ClassCastException} if used on
+ *   the wrong type.
+ * - All setter functions can throw {@link AvroRuntimeException} if used on a field
+ *   which doesn't exist in the schema, which is the same behavior as
+ *   {@link GenericData.Record}.
+ * - All primitive getter functions likewise can throw {@link AvroRuntimeException}
+ *   if used on a field which doesn't exist in the schema.
+ * - All object getter functions can return null if used on a field which doesn't
+ *   exist in the schema, which is the same behavior as {@link GenericData.Record#get()}
+ * - All primitive getter functions can throw a {@link NullPointerException} if used to
+ *   retrieved an optional field for which the value has been set to null. In order
+ *   to avoid this, one can use the {@link #isDefined(String)} function before calling
+ *   a primitive getter.
+ */
+public interface PrimitiveGenericRecord extends GenericRecord {
+  /**
+   * A function to check if a given field contains a non-null value. This can be used
+   * in conjunction with the primitive getters to avoid boxing. For example, with the
+   * regular {@link GenericRecord} API, it is common to do the following:
+   *
+   * Float myField = record.get("myField");
+   * if (myField != null) {
+   *   doSomething(myField);
+   * }
+   *
+   * Instead, with the {@link PrimitiveGenericRecord} API, the above pattern would be
+   * refactored into:
+   *
+   * if (record.isDefined("myField")) {
+   *   doSomething(record.getPrimitiveFloat("myField"));
+   * }
+   *
+   * This allows optional fields to be handled as primitives and thus avoid the boxing
+   * overhead.
+   *
+   * @param key the field name
+   * @return true if the field has a value defined, or false if it is null
+   * @throws AvroRuntimeException
+   */
+  boolean isDefined(String key) throws AvroRuntimeException;
```

Comment:
nit - "isDefined" sounds (to me?) more like if the schema contains a field. maybe hasValue()? of nonNull() ?

Reply:
Yeah, I wasn't sure how to name this one. Also thought of `isSet(name)`... what's your favorite wording?

Reply:
what if its set to null? :-)